File: codes/DequeBaseQueue/C/DequeBaseQueue.c
#include <stdio.h>
#include <stdlib.h>
#include "DequeBaseQueue.h"

void QueueInit(Queue* pq)
{
	DequeInit(&pq->deq);
}

int QIsEmpty(Queue* pq)
{
	return DQIsEmpty(&pq->deq);
}

void Enqueue(Queue* pq, Data data)
{
	DQAddLast(&pq->deq, data);
}

Data Dequeue(Queue* pq)
{
	return DQRemoveFirst(&pq->deq);
}

Data QPeek(Queue* pq)
{
	return DQGetFirst(&pq->deq);
}

File: codes/DequeBaseQueue/Java/DequeBaseQueue/DequeBaseQueue.java
package DequeBaseQueue;

public class DequeBaseQueue<T> implements DequeBaseQueueInterface<T> {
    private Deque<T> deq;

    public DequeBaseQueue() {
        this.deq = new Deque<T>();
    }
    @Override
    public boolean QIsEmpty() {
        return this.deq.DQIsEmpty();
    }

    @Override
    public void Enqueue(T data) {
        this.deq.DQAddLast(data);
    }

    @Override
    public T Dequeue() {
        return this.deq.DQRemoveFirst();
    }

    @Override
    public T QPeek() {
        return this.deq.DQGetFirst();
    }
}
